client/gitlab/client.go

```diff
@@ -0,0 +1,93 @@
+package gitlab
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/mendersoftware/integration-test-runner/logger"
+	"github.com/xanzy/go-gitlab"
+)
+
+// Client represents a GitLab client
+type Client interface {
+	CancelPipelineBuild(path string, id int) error
+	CreatePipeline(path string, options *gitlab.CreatePipelineOptions) (*gitlab.Pipeline, error)
+	GetPipelineVariables(path string, id int) ([]*gitlab.PipelineVariable, error)
+	ListProjectPipelines(path string, options *gitlab.ListProjectPipelinesOptions) (gitlab.PipelineList, error)
+}
+
+type gitLabClient struct {
+	client     *gitlab.Client
+	dryRunMode bool
+}
+
+// NewGitLabClient returns a new GitLabClient for the given conf
+func NewGitLabClient(accessToken string, baseURL string, dryRunMode bool) (Client, error) {
+	gitlabClient := gitlab.NewClient(nil, accessToken)
+	err := gitlabClient.SetBaseURL(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	return &gitLabClient{
+		client:     gitlabClient,
+		dryRunMode: dryRunMode,
+	}, nil
+}
+
+// CancelPipelineBuild cancel a pipeline
+func (c *gitLabClient) CancelPipelineBuild(path string, id int) error {
+	if c.dryRunMode {
+		msg := fmt.Sprintf("gitlab.CancelPipelineBuild: path=%s,id=%d",
+			path, id,
+		)
+		logger.GetRequestLogger().Push(msg)
+		return nil
+	}
+	_, _, err := c.client.Pipelines.CancelPipelineBuild(path, id, nil)
+	return err
+}
+
+// CreatePipeline creates a pipeline
+func (c *gitLabClient) CreatePipeline(path string, options *gitlab.CreatePipelineOptions) (*gitlab.Pipeline, error) {
+	if c.dryRunMode {
+		optionsJSON, _ := json.Marshal(options)
+		msg := fmt.Sprintf("gitlab.CreatePipeline: path=%s,options=%s",
+			path, string(optionsJSON),
+		)
+		logger.GetRequestLogger().Push(msg)
+		return &gitlab.Pipeline{}, nil
+	}
+	pipeline, _, err := c.client.Pipelines.CreatePipeline(path, options, nil)
+	return pipeline, err
+}
+
+// GetPipelineVariables get the pipeline variables
+func (c *gitLabClient) GetPipelineVariables(path string, id int) ([]*gitlab.PipelineVariable, error) {
+	if c.dryRunMode {
+		msg := fmt.Sprintf("gitlab.GetPipelineVariables: path=%s,id=%d",
+			path, id,
+		)
+		logger.GetRequestLogger().Push(msg)
+		return []*gitlab.PipelineVariable{}, nil
+	}
+	variables, _, err := c.client.Pipelines.GetPipelineVariables(path, id, nil)
+	return variables, err
+}
+
+// ListProjectPipelines list the project pipelines
+func (c *gitLabClient) ListProjectPipelines(path string, options *gitlab.ListProjectPipelinesOptions) (gitlab.PipelineList, error) {
+	if c.dryRunMode {
```

Comment:
Musing:
One idea could have been to have a dry-run client instead, and separate it out completely? :thinking:

Reply:
It's doable (and I even thought about that approach), but at the end implementing the dryRunMode circuit breaker was faster.